### ISSUE TITLE
Make moving/scaling/rotating live

### ISSUE THREAD

**Issue by humdingerb** [threedeyes/STLover#7]
...instead of applying the transformations after clicking "OK" in the dialog.

**Comment by Lt-Henry** [comment]
I can give it a try to this ticket (or sequence of tickets). Do you want live transformation to be applied to 3d model? wouldn't be nice to a ghost model in wireframe or flat color to show both after and before transformation? We need some reference here.

**Comment by Lt-Henry** [comment]
![s8](https://user-images.githubusercontent.com/3853371/129494347-f359670f-658a-4090-85fe-fc86657fc000.jpeg)
![s9](https://user-images.githubusercontent.com/3853371/129494346-333d0967-394b-4221-ac42-b07953915a8d.jpeg)


Ok, I have currently this. It is only working for scale, but it should be easily extended to all transform operations. Preview model is updated whenever user sets a new scale. 

Question is... could be this enough as a first solution? How you propose to do that? With just a single model instead of original+preview? I am open to suggestions here, the underlying mechanism should be the same.

**Comment by humdingerb** [comment]
It's of course 3deyes call, but personally, I'd go for adding BSpinner(s) to the input field(s) and updating the view live with the changes. Add a "Revert" button that is SetEnabled(true) when the values have changed that reverts to when the transform window was first opened.

**Comment by threedeyes** [comment]
> It's of course 3deyes call, but personally, I'd go for adding BSpinner(s) to the input field(s) and updating the view live with the changes. Add a "Revert" button that is SetEnabled(true) when the values have changed that reverts to when the transform window was first opened.

+1 )

**Comment by threedeyes** [comment]
> Ok, I have currently this. It is only working for scale, but it should be easily extended to all transform operations. Preview model is updated whenever user sets a new scale.
> 
> Question is... could be this enough as a first solution? How you propose to do that? With just a single model instead of original+preview? I am open to suggestions here, the underlying mechanism should be the same.

How about making the preview model transparent?

**Comment by Lt-Henry** [comment]
> > Ok, I have currently this. It is only working for scale, but it should be easily extended to all transform operations. Preview model is updated whenever user sets a new scale.
> > Question is... could be this enough as a first solution? How you propose to do that? With just a single model instead of original+preview? I am open to suggestions here, the underlying mechanism should be the same.
> 
> How about making the preview model transparent?

I've already thought it. Problem is, transparency triangles must be sorted before pushing to raster, because z-buffer doesn't care about transparent stuff. I can do it, if you like, but it will take a while. I can try to ignore sorting and see what happens, at least as a mock up. Would be nice to have a reference software yo copy from.

**Comment by Lt-Henry** [comment]
> It's of course 3deyes call, but personally, I'd go for adding BSpinner(s) to the input field(s) and updating the view live with the changes. Add a "Revert" button that is SetEnabled(true) when the values have changed that reverts to when the transform window was first opened.

About BSpinner... It makes sense on rotating transforms, as are naturally bounded to [0 , 359] degrees. But what about scale and translate? Perhaps a 1x to 10x could be enough for scale, on 0.1 steps. Translate should be smarter than that.

About revert button. On my current concept it makes no sense, user sets some transform value, validate results with live preview model, and then he can press either Ok, to apply transform or Cancel, so tool window is closed and preview hidden with no changes at all.

I am not imposing any workflow here! I dont have problems implementing whatever workflow we agreed on :D

Edit: BSpinner != BSlider. So yes, Spinners can be used on scale/translate and a slider for rotations!

**Comment by humdingerb** [comment]
WRT the "Revert" button concept, that only makes sense if there were no preview, right. If there's a preview, OK/Cancel buttons are the way to go. The "Revert" concept would make it possible to leave those transform windows open and manipulate in parallel. Might be neat, or not... a real user of the app may have more relevant input than me.

As long as direct value entry is possible, those BSpinner's can easily have a large step size, e.g. scaling in .25 steps etc.

**Comment by Lt-Henry** [comment]
Blender uses your workflow, there is no ghost preview, just live transform that can be either accepted or reverted.

**Comment by Lt-Henry** [comment]
btw, sooner or later we should move to modern OpenGL version. With shaders we can get some nice features, like model outline, or cross section view (handy feature for a 3D print preview).

**Comment by Lt-Henry** [comment]
![Screenshot_20210818_010851](https://user-images.githubusercontent.com/3853371/129812558-7d4ceb97-dda1-410b-af98-dce773575b7d.png)

This is with just a model on screen, but yellow colored while it is being transformed then back to white. Bounding box is not updated: bug or feature? :D

So now, Do I have to change "Cancel" with "Revert" and then reset to defaults without closing tool window, right?
